testutil/ciao-down/README.md

```diff
@@ -10,39 +12,336 @@ Then simply type
+done in the cloud-init file discussed below.
+
+### The Cloudinit document
+
+The third document contains a cloud-init user data file that can be used
+to create users, mount folders, install and update packages and perform
+general configuration tasks on the newly created VM.  The user is
+expected to be familiar with the
+[cloudinit file format](https://cloudinit.readthedocs.io/en/latest/).
+
+Like the instance specification and data documents, the cloudinit document
+is processed by the Go template engine before being passed to cloudinit
+running inside the guest VM.  The template engine makes a number of
+functions available to the cloudinit document.  These functions are used
+to perform common tasks such as configuring proxies and communicating with
+ciao-down running on the host.  Each function must be passed the
+workspace object as the first parameter.  The following functions are
+available:
+
+#### proxyVars
+
+Generates a string containing variable definitions for all the proxy
+variables.  It is useful for prepending to commands inside the runcmd:
+section of the cloudinit document that require proxies to be set, e.g,
+
+```
+- {{proxyvars .}} wget https://storage.googleapis.com/golang/go1.8.linux-amd64.tar.gz" "/tmp/go1.8.linux-amd64.tar.gz
+```
+
+may expand to something like after being processed by the template engine
+
+```
+- https_proxy=http://myproxy.mydomain.com:911 wget https://storage.googleapis.com/golang/go1.8.linux-amd64.tar.gz" "/tmp/go1.8.linux-amd64.tar.gz
 ```
 
-ciao-down will install some needed dependencies on your local PC such
-as qemu and xorriso. It will then download an Ubuntu Cloud Image and
-create a VM based on this image. It will boot the VM and install in that
-VM everything you need to run ciao Single VM, including docker, ceph,
-go, gcc, etc. When ciao-down prepare has finished you can connect to the
-newly created VM with
+#### proxyEnv
 
+Generates proxy definitions in a format that is suitable for the /etc/environment file.
+It should be used as follows:
+
+```
+write_files:
+{{with proxyEnv . 5}}
+ - content: |
+{{.}}
+   path: /etc/environment
+{{end}}
 ```
-$GOPATH/bin/ciao-down connect
+
+#### download
+
+Download should be used to download files from inside the guest.  Download is preferable
+to direclty downloading files inside the cloudinit document using curl or wget for
+example, as the downloaded files are cached by ciao-down.  This means that the next time
+an instance of this workload is created the file will not have to be rertrieved from
+the Internet, resulting in quicker boot times for the new VM.
+
+download takes three parameters
+
+- The workspace object
+- The URL of the file to download
+- The location on the guest where the downloaded file should be stored
+
+An example of its usage is
+
 ```
+- {{download . "https://storage.googleapis.com/golang/go1.8.linux-amd64.tar.gz" "/tmp/go1.8.linux-amd64.tar.gz"}}
```

Comment:
The download command already does this.  When you use download the file is downloaded by the host, placed in ~/.ciao-down/cache and then served up to the guest via HTTP.  The file is then downloaded from the host and save at the path given above on the guest.  The path given here, /tmp/go1.8.linux-amd64.tar.gz is a guest local path and not a host path.   Needless to say, if the file already exists on the host, it doesn't need to be downloaded from the Internet.

The support for host side caching of files is another factor that will complicate multiple instance support.